File: core/infrastructure/src/main/java/com/gonezo/multiplatform/audioextraction/infrastructure/resolution/ResolutionCoordinatorImpl.java
package com.gonezo.multiplatform.audioextraction.infrastructure.resolution;

import com.gonezo.multiplatform.audioextraction.application.ResolutionCoordinator;
import com.gonezo.multiplatform.audioextraction.domain.model.ExtractionContext;
import com.gonezo.multiplatform.audioextraction.domain.model.FieldCandidate;
import com.gonezo.multiplatform.audioextraction.domain.model.ResolvedField;
import com.gonezo.multiplatform.audioextraction.domain.schema.FieldSchema;
import com.gonezo.multiplatform.audioextraction.domain.schema.OutputSchema;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

public final class ResolutionCoordinatorImpl implements ResolutionCoordinator {
  private final List<FieldResolver> resolvers = List.of(
    new EnumResolver(),
    new DateResolver(),
    new NumberResolver(),
    new StringResolver()
  );

  @Override
  public Map<String, ResolvedField> resolve(
    Map<String, List<FieldCandidate>> candidates,
    OutputSchema schema,
    ExtractionContext context
  ) {
    Map<String, ResolvedField> resolved = new LinkedHashMap<>();

    for (Map.Entry<String, FieldSchema> fieldEntry : schema.fields().entrySet()) {
      String fieldName = fieldEntry.getKey();
      FieldSchema fieldSchema = fieldEntry.getValue();
      List<FieldCandidate> fieldCandidates = candidates.getOrDefault(fieldName, List.of());

      ResolvedField field = resolveField(fieldName, fieldSchema, fieldCandidates);
      resolved.put(fieldName, normalizeIssues(field));
    }

    return resolved;
  }

  private ResolvedField resolveField(String fieldName, FieldSchema schema, List<FieldCandidate> candidates) {
    for (FieldResolver resolver : resolvers) {
      if (resolver.supports(schema)) {
        return resolver.resolve(fieldName, schema, candidates);
      }
    }

    if (schema.required()) {
      return new ResolvedField(null, 0D, List.of(), List.of("missing"));
    }

    return new ResolvedField(null, 0D, List.of(), List.of());
  }

  private ResolvedField normalizeIssues(ResolvedField field) {
    if (field.issues().isEmpty()) {
      return field;
    }

    List<String> normalizedIssues = new ArrayList<>();
    for (String issue : field.issues()) {
      String value = issue == null ? "" : issue.trim();
      if ("missing".equals(value) || "ambiguous".equals(value) || "invalid".equals(value) || "invalid_format".equals(value)) {
        normalizedIssues.add(value);
      } else {
        normalizedIssues.add("invalid");
      }
    }

    return new ResolvedField(field.value(), field.confidence(), field.evidence(), normalizedIssues);
  }
}

File: core/src/main/kotlin/com/gonezo/audioextraction/infrastructure/resolution/ResolutionCoordinatorImpl.kt
package com.gonezo.audioextraction.infrastructure.resolution

import com.gonezo.audioextraction.application.pipeline.ResolutionCoordinator
import com.gonezo.audioextraction.domain.model.ExtractionContext
import com.gonezo.audioextraction.domain.model.FieldCandidate
import com.gonezo.audioextraction.domain.model.ResolvedField
import com.gonezo.audioextraction.domain.schema.FieldSchema
import com.gonezo.audioextraction.domain.schema.OutputSchema

class ResolutionCoordinatorImpl : ResolutionCoordinator {
  private val resolvers: List<FieldResolver> = listOf(
    EnumResolver(),
    DateResolver(),
    NumberResolver(),
    StringResolver(),
  )

  override fun resolve(
    candidates: Map<String, List<FieldCandidate>>,
    schema: OutputSchema,
    context: ExtractionContext,
  ): Map<String, ResolvedField> {
    val resolved = linkedMapOf<String, ResolvedField>()

    for ((fieldName, fieldSchema) in schema.fields) {
      val fieldCandidates = candidates[fieldName] ?: emptyList()
      val field = resolveField(fieldName, fieldSchema, fieldCandidates)
      resolved[fieldName] = normalizeIssues(field)
    }

    return resolved
  }

  private fun resolveField(fieldName: String, schema: FieldSchema, candidates: List<FieldCandidate>): ResolvedField {
    for (resolver in resolvers) {
      if (resolver.supports(schema)) {
        return resolver.resolve(fieldName, schema, candidates)
      }
    }

    return if (schema.required) {
      ResolvedField(null, 0.0, emptyList(), listOf("missing"))
    } else {
      ResolvedField(null, 0.0, emptyList(), emptyList())
    }
  }

  private fun normalizeIssues(field: ResolvedField): ResolvedField {
    if (field.issues.isEmpty()) return field

    val normalizedIssues = field.issues.map { issue ->
      when (issue.trim()) {
        "missing", "ambiguous", "invalid", "invalid_format" -> issue.trim()
        else -> "invalid"
      }
    }

    return field.copy(issues = normalizedIssues)
  }
}
